README.md
<p align="center">
  <a href="https://github.com/vortesnail/qier-progress">
    <img width="271" src="https://cdn.nlark.com/yuque/0/2020/png/341314/1581180114750-bcd1cc60-0847-49e0-96bd-da4e76901f87.png#align=left&display=inline&height=24&name=%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202020-02-09%2000.39.47.png&originHeight=226&originWidth=2560&size=340990&status=done&style=none&width=271">
  </a>
</p>

<h1 align="center">Slim progress bars for anywhere you want to use</h1>
<div align="center">

[![npm version](https://img.shields.io/npm/v/qier-progress)](https://www.npmjs.com/package/qier-progress) [![package size](https://img.shields.io/bundlephobia/minzip/qier-progress)](https://www.npmjs.com/package/qier-progress) 
[![download](https://img.shields.io/npm/dw/qier-progress)](https://www.npmjs.com/package/qier-progress) [![Build Status](https://travis-ci.org/vortesnail/qier-progress.svg?branch=master)](https://travis-ci.org/vortesnail/qier-progress) [![Coverage Status](https://coveralls.io/repos/github/vortesnail/qier-progress/badge.svg?branch=master)](https://coveralls.io/github/vortesnail/qier-progress?branch=master) [![LICENSE](https://img.shields.io/npm/l/qier-progress)](https://github.com/vortesnail/qier-progress/blob/master/LICENSE) 


[简体中文](./READM-zh-CN.md) &#124; English

</div>

## Introduction

`qier-progress` is a progress bar. It can be used for some watting time like jump links, request data, and load or upload files and images to give us feedback and reduce our anxiety. Also if you have used [nprogress](https://github.com/rstacruz/nprogress), then you must know what I am talking about ~

💃[check demo](https://vortesnail.github.io/qier-progress/)

## Quick Start
#### 🛠 Install
```javascript
npm install --save qier-progress
```

#### 📦 Use
Firstly, import module in Vue, React, Angular wherever es6 module is supported .
```javascript
import QProgress from 'qier-progress'
```

Secondly, create instance.
```javascript
const qprogress = new QProgress()
```

Thirdly, simply call `start()` and `finish()` to control the progress bar.
```javascript
qprogress.start()
qprogress.finish()
```

## Advanced usage
#### 📌 Set progress value:
Use `.set(n)` to set a progress percentage, where ![](https://cdn.nlark.com/yuque/__latex/7b8b965ad4bca0e41ab51de7b31363a1.svg#card=math&code=n&height=12&width=10) is a number between `0..1` .
```javascript
qprogress.set(0.0)     // Same as .start()
qprogress.set(0.6)
qprogress.set(1.0)     // Same as .done()
```

#### 🎢 Increase manually:
Use `.inc(n)` to increment the progress bar, but it will stop increasing after reaching the threshold, means it will never reach `100%` .
```javascript
qprogress.inc()
qprogress.inc(0.2)	// specific value you want
```

#### 🥣 Forced finished:
Use `.finish()` to unmount the progress var, of course, there will also have an end process animation.
```javascript
qprogress.finish()
```

#### 🧮 Get current progress value:
Use `.status` to get current value ![](https://cdn.nlark.com/yuque/__latex/7b8b965ad4bca0e41ab51de7b31363a1.svg#card=math&code=n&height=12&width=10) where is a number between `0..1` .
```javascript
qprogress.status
```

## Configuration
#### 🤔 How to customize
When creating an instance, you can customize some parameters like this：
```javascript
const qprogress = new QProgress({
  minimum: 0.08,
  height: 3,
  color: '#17829f'
})
```

#### 📕 Configuration list
| Parameter | Description | Type | Default |
| --- | --- | --- | --- |
| `minimum` | Minimum percentage used upon starting. | number(0..1) | 0.12 |
| `height` | Progress bar's height, unit is `px` . | number | 2 |
| `color` | Progress bar's color, support RGB. | string | '#1890ff' |
| `colorful` | Colorful mode switch. | boolean | true |
| `easing` | Css transition property `time-function` . | string | 'ease' |
| `speed` | Css transition property `duration` , unit is `ms` . | number | 400 |
| `trickle` | Automatic incrementing behavior switch. | boolean | true |
| `trickleSpeed` | Automatic incrementing speed, means increment interval, unit is `ms` . | number | 400 |
| `parentNode` | Specify this to change the parent container. | Element &#124; string | 'body' |


## Contribution
Welcome to participate in this project, please read [CONTRIBUTING](ssd) carefully.

## Inspiration and purpose
First of all, I am a beginner of `typescript` . When I enjoy the convenience brought by [nprogress](https://github.com/rstacruz/nprogress), I hope that I can learn a little bit from it, so I retyped this plugin using `typescript`  and added some other features. I learned a lot of coding knowledge in the process, and finally I sincerely thank the [nprogress contributors](https://github.com/rstacruz/nprogress/graphs/contributors) very much, respect!

## About me
[![Github.svg](https://cdn.nlark.com/yuque/0/2020/svg/341314/1581188387396-7788bf71-e189-4c34-bcaa-eaa5b0055497.svg#align=left&display=inline&height=24&name=Github.svg&originHeight=32&originWidth=32&size=2534&status=done&style=none&width=24)](https://github.com/vortesnail) [![juejin-02.svg](https://cdn.nlark.com/yuque/0/2020/svg/341314/1581188386963-d8bc6ee6-b3f4-47f8-b53c-5bd493c890b4.svg#align=left&display=inline&height=24&name=juejin-02.svg&originHeight=32&originWidth=32&size=884&status=done&style=none&width=24)](https://juejin.im/user/5da573d3f265da5b8a5168a6) [![哔哩哔哩.svg](https://cdn.nlark.com/yuque/0/2020/svg/341314/1581188388001-39360fed-b53a-47db-8b83-cff8126561da.svg#align=left&display=inline&height=24&name=%E5%93%94%E5%93%A9%E5%93%94%E5%93%A9.svg&originHeight=32&originWidth=32&size=6080&status=done&style=none&width=24)](https://space.bilibili.com/80755916) [![知乎.svg](https://cdn.nlark.com/yuque/0/2020/svg/341314/1581188354814-5e2956ab-2895-4f09-b788-7c5175160e41.svg#align=left&display=inline&height=24&name=%E7%9F%A5%E4%B9%8E.svg&originHeight=32&originWidth=32&size=2265&status=done&style=none&width=24)](https://www.zhihu.com/people/vortesnail)

## License
[LICENSE](./LICENSE)

CHANGELOG.md
## [1.0.1](https://github.com/vortesnail/qier-progress/compare/v1.0.0...v1.0.1) (2020-02-09)



# [1.0.0](https://github.com/vortesnail/qier-progress/compare/v0.0.2...v1.0.0) (2020-02-09)



## [0.0.2](https://github.com/vortesnail/qier-progress/compare/9665671be92abbf2b26c4782a38f1183184ba292...v0.0.2) (2020-02-09)


### Features

* basic construct like default configure ([d0b2852](https://github.com/vortesnail/qier-progress/commit/d0b285213122155fece5428102eb370c2a539043))
* Complete function development completed ([1f89df5](https://github.com/vortesnail/qier-progress/commit/1f89df516682c7ddc056ddbb550bb9d3c32a1344))
* initial project ([9665671](https://github.com/vortesnail/qier-progress/commit/9665671be92abbf2b26c4782a38f1183184ba292))





src/index.ts
/**
 *  Author： vortesnail
 *  License： MIT
 *  Github： https://github.com/vortesnail
 */
import { Configuration } from './types'
import {
  middleNum,
  addClass,
  toBarPerc,
  css,
  queue,
  getPositioningCSS,
  barPositionCSS,
  remove
} from './helpers'
import './index.css'

/********************************************************
 *  Usage:                                              *
 *    const qprogress = new QProgress({ minimum: 0.1 }) *
 *    qprogress.start()                                 *
 *    // some code...                                   *
 *    qprogress.finish()                                *
 ********************************************************/

class QProgress {
  // Configuration
  private config: Configuration
  // Last number
  public status: number | null = null

  constructor(config?: Configuration) {
    this.config = {
      minimum: 0.12,
      height: 2,
      color: '#1890ff',
      colorful: true,
      easing: 'ease',
      speed: 400,
      positionUsing: '',
      trickle: true,
      trickleSpeed: 400,
      parentNode: 'body',
      barSelector: '[role="bar"]',
      pegSelector: '[role="peg"]',
      template: '<div class="bar" role="bar"><div class="peg" role="peg"></div></div>',
      ...config
    }
  }

  /**
   * Sets the progress bar status, where `n` is a number from `0.0` to `1.0`.
   *
   *     qprogress.set(0.4)
   *     qprogress.set(1.0)
   */

  public set(n: number): void {
    const started: boolean = this.isStarted()

    n = middleNum(n, this.config.minimum!, 1)
    this.status = n === 1 ? null : n

    const { barSelector, speed: speedCopy, easing, parentNode } = this.config
    let { positionUsing } = this.config

    let progress = this.render(!started)
    const bar: HTMLElement | null = progress.querySelector(barSelector!)
    const speed: number = speedCopy!
    const ease: string = easing!

    progress.offsetWidth /* Compulsory repainting */

    queue(function(next: Function) {
      // Set positionUsing if it hasn't already been set
      if (positionUsing === '') positionUsing = getPositioningCSS()

      // Add transition
      css(bar!, barPositionCSS(n, speed, ease, positionUsing!))

      if (n === 1) {
        // Fade out
        css(progress, {
          transition: 'none',
          opacity: 1
        })

        progress.offsetWidth /* Compulsory repainting */

        setTimeout(function() {
          css(progress, {
            transition: `all ${speed}ms linear`,
            opacity: 0
          })
          setTimeout(function() {
            remove(parentNode as HTMLElement)
            next()
          }, speed)
        }, speed)
      } else {
        setTimeout(next, speed)
      }
    })
  }

  /**
   * Checks if the progress is started, `0.0` to `1.0`: `true`, `null`: `false` .
   */

  private isStarted(): boolean {
    return typeof this.status === 'number'
  }

  /**
   * Shows the progress bar.
   * This is the same as setting the status to 0%, except that it doesn't go backwards.
   *
   *     qprogress.start()
   *
   */
  public start(): void {
    const _this = this
    const { trickle, trickleSpeed } = this.config
    if (!this.status) this.set(0)

    const work = function() {
      setTimeout(function() {
        if (!_this.status) return
        _this.trickle()
        work()
      }, trickleSpeed)
    }

    if (trickle) work()
  }

  /**
   * Hides the progress bar.
   * This is the *sort of* the same as setting the status to 100%, with the
   * difference being `finish()` makes some placebo effect of some realistic motion.
   *
   *     qprogress.finish()
   *
   * If `true` is passed, it will show the progress bar even if its hidden.
   *
   *     qprogress.finish(true)
   */

  public finish(force?: boolean): void {
    if (!force && !this.status) return

    this.inc(0.3 + 0.5 * Math.random())
    this.set(1)
  }

  private trickle(): void {
    this.inc()
  }

  /**
   * Increments by a random amount.
   */

  public inc(amount?: number) {
    let n: number | null = this.status

    if (!n) {
      return this.start()
    } else if (n > 1) {
      return
    } else {
      if (typeof amount !== 'number') {
        if (n >= 0 && n < 0.2) {
          amount = 0.1
        } else if (n >= 0.2 && n < 0.5) {
          amount = 0.04
        } else if (n >= 0.5 && n < 0.8) {
          amount = 0.02
        } else if (n >= 0.8 && n < 0.99) {
          amount = 0.005
        } else {
          amount = 0
        }
      }

      n = middleNum(n + amount, 0, 0.994)
      return this.set(n)
    }
  }

  /**
   * Checks if the progress bar is rendered.
   */

  private isRendered(): boolean {
    return !!document.getElementById('qprogress')
  }

  /**
   * renders the progress bar markup based on the `template` setting.
   */

  private render(fromStart: boolean): HTMLElement {
    if (this.isRendered()) return document.getElementById('qprogress')!

    addClass(document.documentElement, 'qprogress-busy')

    const { template, barSelector, pegSelector, parentNode, height, color, colorful } = this.config

    const progress: HTMLElement = document.createElement('div')
    progress.id = 'qprogress'
    progress.innerHTML = template!

    const bar: HTMLElement | null = progress.querySelector(barSelector!)
    const peg: HTMLElement | null = bar!.querySelector(pegSelector!)
    const perc: string = fromStart ? '-100' : toBarPerc(this.status || 0)
    const parent: HTMLElement =
      typeof parentNode === 'string'
        ? (document.querySelector(parentNode) as HTMLElement)
        : (parentNode as HTMLElement)

    css(bar!, {
      height: `${height}px`,
      backgroundColor: color,
      transition: 'all 0 linear',
      transform: `translate3d(${perc}%, 0, 0)`
    })

    if (!colorful) {
      css(peg!, {
        boxShadow: `0 0 10px ${color}, 0 0 5px ${color}`
      })
    } else {
      bar && addClass(bar, 'colorful-bar')
    }

    if (parent !== document.body) {
      addClass(parent, 'qprogress-custom-parent')
    }

    parent.appendChild(progress)
    return progress
  }
}

export default QProgress
